components/region_cache_memory_engine/src/range_stats.rs

```diff
@@ -0,0 +1,441 @@
+        self.num_regions.load(Ordering::Relaxed)
+    }
+
+    /// Collect candidates for eviction sorted by activity in creasing order:
+    ///
+    /// 1. Get all the regions sorted (decreasing) by region activity using
+    ///    [raftstore::coprocessor::RegionCollector::handle_get_top_regions].
+    /// 2. Remove all regions where `is_cached_pred` returns false when passed
+    ///    the region's range or those that have been loaded for less than
+    ///    `self.evict_min_duration`.
+    /// 3. Reverse the list so that it is now sorted in the order of increasing
+    ///    activity.
+    /// 4. Store the results in `ranges_out` using [Vec::extend].
+    pub fn collect_candidates_for_eviction<F>(
+        &self,
+        ranges_out: &mut Vec<(CacheRange, u64)>,
+        is_cached_pred: F,
+    ) where
+        F: Fn(&CacheRange) -> bool,
+    {
+        // Gets all of the regions, sorted by activity.
+        let all_regions = self.info_provider.get_top_regions(None).unwrap();
+        let regions_loaded = self.region_loaded_at.read().unwrap();
+        ranges_out.extend(
+            all_regions
+                .iter()
+                .filter_map(|(region, approx_size)| {
+                    let r = CacheRange::from_region(region);
+                    is_cached_pred(&r)
+                        .then(|| {
+                            match regions_loaded.get(&region.get_id()) {
+                                // Do not evict ranges that were loaded less than
+                                // `EVICT_MIN_DURATION` ago.
+                                Some(&time_loaded)
+                                    if Instant::now() - time_loaded < self.evict_min_duration =>
+                                {
+                                    None
+                                }
+                                Some(_) | None =>
+                                // None indicates range loaded from a hint, not by this manager.
+                                {
+                                    Some((r, *approx_size))
+                                }
+                            }
+                        })
+                        .flatten()
+                })
+                .rev(),
+        );
+    }
+
+    /// This method should be called when `evicted_range` is succesfully evicted
+    /// to remove any internal `RegionStatsManager` that corresponds to the
+    /// range.
+    ///
+    /// Calls [raftstore::coprocessor::region_info_accessor::RegionInfoProvider::find_region_by_key] to
+    /// find the region corresponding to the range.
+    ///
+    /// TODO (afeinberg): This is inefficient, either make this method bulk, or
+    /// find another way to avoid calling `find_region_by_key` in a loop.
+    pub fn handle_range_evicted(&self, evicted_range: &CacheRange) {
+        // TODO (afeinberg): This is inefficient.
+        let _ = self
+            .info_provider
+            .find_region_by_key(&evicted_range.start)
```

Comment:
we should not use evicted_range.start directly here as the key in CacheRange contains the DATA_PREFIX, where as key in region does not. We should use evicted_range.start[1..] instead. @afeinberg